### PR TITLE
Allow costs and window size limits to be specified in the config_list

### LINKED ISSUE #1680: [Feature Request]: Generalize token accounting and budgeting
### Is your feature request related to a problem? Please describe.

At present, model token limits are hardcoded here:
https://github.com/microsoft/autogen/blob/f68c09b035219b9d28d7f2999d1e971e6243a8df/autogen/token_count_utils.py#L16-L34

Token costs are hardcoded here:
https://github.com/microsoft/autogen/blob/f68c09b035219b9d28d7f2999d1e971e6243a8df/autogen/oai/openai_utils.py#L23-L53

However, there are a number of issues with this including:

- The name "max_token_limit" is confusing and can clash with "max_tokens", which is a parameter of the chat completion request, and refers exclusively to output tokens.
- Adding new models requires editing both files. It's easy to forget one or the other.
- It's keyed off model names, but those name may not be reliable. For example, "gpt-4" on Azure might point to an 8k, 32k, or 128k model. There is no requirement that these match with OpenAI names. Likewise, model names may not match at all (e.g., "my-gpt-4")
- Adding new models or providers requires hand-editing core library files.

### Describe the solution you'd like

I would like to be able to provide token accounting information in the OAI_CONFIG_LIST.

Perhaps something like this:

```
[
        "model": "gpt-4-turbo",
        "api_key": "blahblahblah",
        "base_url": "https://mymodel.openai.azure.com/",
        "api_type": "azure",
        "api_version": "2023-12-01-preview",
        "window_size": (128000, 4096),
        "1k_token_cost": (0.01, 0.03) 
]
```

In the case where input and output tokens are not distinguished:
```
[
        "model": "gpt-3.5-turbo",
        "api_key": "blahblahblah",
        "base_url": "https://mymodel.openai.azure.com/",
        "api_type": "azure",
        "api_version": "2023-12-01-preview",
        "window_size": 16385,
        "1k_token_cost": (0.0005, 0.0015) 
]
```

We would then use these values when present to compute costs, token limits, etc. 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by afourney**
## Why are these changes needed?

See issue #1680.

TL;DR: 
Token limits and costs are hardcoded in various files. Adding new models requires editing numerous files. Costs are keyed off model names, but those name may not be reliable. For example, "gpt-4" on Azure might point to an 8k, 32k, or 128k model. There is no requirement that these match with OpenAI names. Likewise, model names may not match at all (e.g., "my-gpt-4")
Adding new models or providers requires hand-editing core library files.

This PR allows costs and limits to be specified in the config. E.g., 

```
[
        "model": "gpt-4-turbo",
        "api_key": "blahblahblah",
        "base_url": "https://mymodel.openai.azure.com/",
        "api_type": "azure",
        "api_version": "2023-12-01-preview",
        "window_size":  { "input": 128000, "output":  4096 },
        "token_cost_1k":  { "input": 0.01, "output": 0.03 ) 
]
```

Limitations: Vision models are not fully-supported. We can compute the cost only after we get the tokens counts back after making the call.

## Related issue number

Closes #1680, #1270

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: autogen/oai/client.py [hunk 1]
```diff
@@ -314,6 +327,8 @@ class OpenAIWrapper:
         "api_version",
         "api_type",
         "tags",
+        "window_size",
```
Comment (olgavrou) [inline autogen/oai/client.py:330]: are we doing anything with window size?

**Review by olgavrou — COMMENTED** [review]
wondering if this needs to be passed into `cost` or if we can just pass all the args into the client constructor (as I believe is already done?) and re-use in different methods, that way we don't have to change the API every time we want to use a new config value

**Comment by afourney** [comment]
> wondering if this needs to be passed into `cost` or if we can just pass all the args into the client constructor (as I believe is already done?) and re-use in different methods, that way we don't have to change the API every time we want to use a new config value

I'd love to figure this out. I think the issue I was working around is that the main wrapper client takes the config list, which may create many OAI or Azure clients, and we don't know which is used until we make the call. The cost method I had initially targeted for modification is a member of the wrapper client class. 

I'm not sure I fully understand the solution to this you are proposing, but I would be happy to get rid of the new parameter if possible.

File: autogen/oai/client.py (hunk 1 above)
Comment (afourney) [reply autogen/oai/client.py:330]: Yes. Both :)

I have yet to modify WebSurfer (this is still a raft PR), but it will use WindowSize to understand how to break up pages. 

I also use it in the complex_tasks branch with SocietyOfMind and will port those changes to main once this is ready and merged.

**Comment by olgavrou** [comment]
> > wondering if this needs to be passed into `cost` or if we can just pass all the args into the client constructor (as I believe is already done?) and re-use in different methods, that way we don't have to change the API every time we want to use a new config value
> 
> I'd love to figure this out. I think the issue I was working around is that the main wrapper client takes the config list, which may create many OAI or Azure clients, and we don't know which is used until we make the call. The cost method I had initially targeted for modification is a member of the wrapper client class.
> 
> I'm not sure I fully understand the solution to this you are proposing, but I would be happy to get rid of the new parameter if possible.

If we pass any extra kwargs that we need to the client constructors so for example `OpenAIClient`, then when `OpenAIClient`'s `cost` or `create` or other function is called it will have access to things like token cost, window size, etc

File: autogen/oai/client.py [hunk 2]
```diff
@@ -275,21 +277,32 @@ def create(self, params: Dict[str, Any]) -> ChatCompletion:
 
         return response
 
-    def cost(self, response: Union[ChatCompletion, Completion]) -> float:
+    def cost(
+        self,
+        response: Union[ChatCompletion, Completion],
+        token_cost_1k: Optional[Union[float, Dict[str, float]]] = None,
+    ) -> float:
```
Comment (maxim-saplin) [inline autogen/oai/client.py:284]: Not critical, yet it might be reasonable to switch to Decimal type (it smells to have float for currency/money values :) - openai_utils.py included

**Review by maxim-saplin — COMMENTED** [review]
Looks good. Though I think docs is an important piece that is missing from this PR:
- https://microsoft.github.io/autogen/docs/llm_configuration
- Docstrings at `openai_utils.py`

**Comment by kevin666aa** [comment]
Should we force or at least warn user to pass in token_cost if they are using azure?

From my previous experience azure always return 'gpt-4' in response['model'], this will very likely cause a wrong calculation of cost.
See #1270

**Comment by afourney** [comment]
> Should we force or at least warn user to pass in token_cost if they are using azure?
> 
> From my previous experience azure always return 'gpt-4' in response['model'], this will very likely cause a wrong calculation of cost. See #1270

That is a good idea. This has been my experience too.  Perhaps it should warn or force for any model that isn't OpenAI.

**Comment by afourney** [comment]
Closing this PR, since it is quite dated. Will open a new issue.
